library/core/src/str/traits.rs

```diff
@@ -194,19 +195,53 @@ unsafe impl const SliceIndex<str> for ops::Range<usize> {
         }
     }
     #[inline]
-    unsafe fn get_unchecked(self, slice: *const str) -> *const Self::Output {
-        let slice = slice as *const [u8];
+    unsafe fn get_unchecked(self, string: *const str) -> *const Self::Output {
+        let slice = string as *const [u8];
         // SAFETY: the caller guarantees that `self` is in bounds of `slice`
         // which satisfies all the conditions for `add`.
-        let ptr = unsafe { slice.as_ptr().add(self.start) };
+        let ptr = unsafe {
+            let this = ops::Range { ..self };
+            assert_unsafe_precondition!(
+                "str::get_unchecked requires that the range is within the string slice",
+                (this: ops::Range<usize>, slice: *const [u8]) =>
+                this.end >= this.start && this.end <= slice.len()
+            );
+            let this = ops::Range { ..self };
+            assert_unsafe_precondition!(
+                "str::get_unchecked requires that the range lies on char boundaries",
+                (this: ops::Range<usize>, string: *const str) =>
+                // SAFETY: the caller guarantees that the `string` pointer is valid
+                unsafe {
+                    (&*string).is_char_boundary(this.start) && (&*string).is_char_boundary(this.end)
```

Comment:
`string` here is a `*const str`. It may be aliased in creative ways. Reborrowing into `&str` may add UB to the program, and in fact doing any read at all may be problematic. We might be able to salvage this situation by hoisting the check all the way out to the actual definition of `str::get_unchecked`, which is in a different file. There we have a `&str` not a `*const str`, so we can do reads. (exactly how to implement this check out there, I am not sure at the moment)

cc @RalfJung to correct me if I'm wrong here

Reply:
Hard to say without knowing how this function is used. But yeah this changes the function to conceptually do a read of the entire string, which it did not do before.

Reply:
For reference, the docs for `SliceIndex::get_unchecked`:

> Returns a shared reference to the output at this location, without
> performing any bounds checking.
>  Calling this method with an out-of-bounds index or a dangling `slice` pointer
> is *undefined behavior* even if the resulting reference is not used.

Reply:
There is no restriction on aliasing, so that could be a problem. Reading should be fine - I could change `is_char_boundary` to just use pointers, but it's probably not worth it. Better to move those checks up the line instead.

Looking at the docs, it seems I should probably add a check that the pointer is not null or dangling.

Reply:
Unfortunately you can't check if a pointer is dangling without a runtime/interpreter like Miri.

Reading through a pointer has aliasing implications: The read means that any `&mut` derived from that pointer is no longer unique. And that is problematic if we are doing optimizations that assume `&mut` is unique. (In Stacked Borrows, a read access disables and Unique tags above it, which makes using those Unique tags later flagged as UB).

Just trying to make sure you understand my thoughts here, I see we agree that moving these checks up is better 👍

Reply:
I'm having some trouble moving these checks up to the `str::get_unchecked(_mut)` level. Those fns are generic for any `SliceIndex`, but there's no way to get the bounds in that trait. I don't think it's worth it to add that capability just for this.

There are hacky ways around that, like using `pointer::offset_from` on the output, but they're not pretty. And they would still require changing the `SliceIndex<str>` bound to `SliceIndex<str, Output = str>`.

Reply:
If you cast the `&str` to a `*const u8`, can you apply the `SliceIndex`, then basically work off the implementation of `str::is_char_boundary` to check the start and end? The implementation of that function is byte-oriented, just the API isn't. We have helpers for other assertions in `core/src/intrinsics.rs` already, you could factor out another helper to keep the additional code down.

Reply:
You should be able to use `&[u8]` not `*const u8`. It would probably be a lot easier.

Reply:
> If you cast the `&str` to a `*const u8`, can you apply the `SliceIndex`, then basically work off the implementation of `str::is_char_boundary` to check the start and end?

I think that would require adding a `SliceIndex<[u8], Output = [u8]>` bound, which would be a breaking change, would it not?

Reply:
Aaaa `SliceIndex` doesn't return `Self`. I temporarily forgot that.

`SliceIndex` is a sealed trait so it is _possible_ we can do something clever by extending it. But I think fiddling with publicly visible APIs is a step too far.

Surely checking this isn't impossible, but an acceptable way to pull this off might just be beyond me. Maybe Mara or someone else I bother will have a suggestion.